fix pass verbose to fit in historical_forecast

turbotimon opened the pull request:
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

<!-- Please mention an issue this pull request addresses. -->
Fixes #2804.

### Summary

<!-- Provide a general description of the code changes in your pull
request. If your pull request is not ready to merge, please create
a draft and ask for comments. -->

verbose now get passed from historical_forecast onto fit (and not only predict) so progressbar etc are not shown during retrain

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, code samples, or others. -->

<!--Thank you for contributing to darts! -->

dennisbader reviewed (changes requested):
Thanks for giving this a go @turbotimon 🚀 . This fails because not all models support the `verbose` parameter in `fit()`.
We would have to change the API of `ForecastingModel.fit()` and subclasses to accept the `verbose` param even if it is not used.

On another thought, I think it could be valuable to make the `verbose` param of `historical_forecasts()` customizable to which level it should act on. For example, I assume that many users are interested in the global progress of the historical forecasts but not in the progress bars of fit and predict.

Basically verbose could accept either `bool` (show everything or nothing) and another type that specifies the level. Maybe an int in `{-1, 0, 1}` with `-1` meaning show nothing, `0` means show only historical forecasts progress, and `1` means show everything). Just an idea. Let me know what you think :)

dennisbader commented:
Alright, then go ahead with solution 1 :)

turbotimon commented:
@dennisbader I added solution 1 with the following comments:

- `verbose` is currently not passed to `super().fit` in most cases, because it also not done in most [`predict`](https://github.com/turbotimon/darts/blob/fix-hfcs-verbose-2804/darts/models/forecasting/exponential_smoothing.py#L146)
- `verbose` is only passed to the underlying `self.model.fit()` if the underlying model supports it
- a hack is needed for `CatBoostModel`, because it derived from `SKLearnModel` but SKLearnModel normally do not support `verbose` in their `fit` method (see code  changes for details)
- I added `verbose=False` to a couple of test to proof if it can be passed without errors. I opted for not implementing additional test just for that to not increase test time.

What do you think?

dennisbader commented on darts/models/forecasting/baselines.py:
```diff
@@ -328,6 +328,7 @@ def fit(
         past_covariates: Optional[Union[TimeSeries, Sequence[TimeSeries]]] = None,
         future_covariates: Optional[Union[TimeSeries, Sequence[TimeSeries]]] = None,
         sample_weight: Optional[Union[TimeSeries, Sequence[TimeSeries], str]] = None,
+        verbose: bool = False,
```
propagate it to the _fit_wrapper of forecasting models (further below)

dennisbader commented on darts/models/forecasting/xgboost.py:
```diff
@@ -278,7 +278,7 @@ def fit(
             Same as for `sample_weight` but for the evaluation dataset.
         **kwargs
-            Additional kwargs passed to `xgb.XGBRegressor.fit()`
+            Additional kwargs passed to `xgb.XGBRegressor.fit()`, e.g. `verbose`.
```
let's make verbose an explicit parameter like in the other models

dennisbader commented on darts/tests/models/forecasting/test_sklearn_models.py:
```diff
@@ -1098,7 +1098,9 @@ def test_fit(self, config):
         with pytest.raises(ValueError):
             model_instance = model(lags=4, lags_past_covariates=4, multi_models=mode)
-            model_instance.fit(series=series, past_covariates=sine_multivariate1)
+            model_instance.fit(
```
let's revert this change, since we want to check for a specific error, and verbose should not be part of it

dennisbader commented on CHANGELOG.md:
```diff
@@ -37,6 +37,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
 
 - Fixed some issues in `NLinearModel` with `normalize=True` that resulted in decreased predictive accuracy. Using `shared_weights=True` and auto-regressive forecasting now work properly. [#2757](https://github.com/unit8co/darts/pull/2757) by [Timon Erhart](https://github.com/turbotimon).
 - Fixed a bug when training a `TorchForecastingModel`, where using certain `torchmetrics` that require a 2D model output (e.g. R2Score) raised an error. [He Weilin](https://github.com/cnhwl).
+- Fixed a bug in `historical_forecast`: Argument `verbose` now gets passed `fit` during retrain. [#2805](https://github.com/unit8co/darts/pull/2805) by [Timon Erhart](https://github.com/turbotimon).
```
Some minor rephrasing. Could you also add the following line to bottom of the "Improved" section?

```
- Added parameter `verbose` to `ForecastingModel.fit()` that allows to control the verbosity for model fitting. Ignored if the underlying model does not support it. [#2805](https://github.com/unit8co/darts/pull/2805) by [Timon Erhart](https://github.com/turbotimon).
```


```suggestion
- Fixed a bug in `historical_forecast` where the `verbose` parameter was not propagated to the the model's `fit()` method. [#2805](https://github.com/unit8co/darts/pull/2805) by [Timon Erhart](https://github.com/turbotimon).
```

dennisbader reviewed (changes requested):
Thanks for the updates @turbotimon, it looks great now 🚀 

Just had a couple minor suggestions, after that it's ready to be merged :)

dennisbader reviewed (changes requested):
Thanks for the great updates @turbotimon and also sorry for the waiting - I was on holidays :) Everything looks good. 

I quickly checked the new behavior for a TorchForecastingModel and now I have to admit that your solution 3. from a while ago seems like the preferred option.

> 3. But another option could be to allow passing verbose explicit in fit_kwargs and predict_kwargs and it would override the verbosity received from historical_forecast for these methods. This would basically solve 2. but with more flexibility (e.g. only fit but not predict). But this solution would presume 1. 

I also think that it should be possible to deactivate the fit progress bar while still showing the historical forecast progress (as you mentioned in the original issue).

Would it be okay for you to apply this suggestion? I could also do the changes if you prefer that :) 

Sorry for the back and forth. I hope it doesn't inflict too much extra effort 🙏

dennisbader commented:
@turbotimon perfext, and I agree with keeping the params in the API.

Sounds good, I'll take it over. Thanks again for the great work!
